File: apps/backend/app/api/routes/scripts.py
from fastapi import APIRouter, Depends, Query, Response, status
from fastapi.responses import JSONResponse, PlainTextResponse
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.script import (
    ScriptCreate,
    ScriptGenerateRequest,
    ScriptGenerateResponse,
    ScriptRead,
    ScriptUpdate,
)
from app.services.script_service import ScriptService

router = APIRouter(tags=["scripts"])


@router.get("/projects/{project_id}/scripts", response_model=list[ScriptRead])
def list_scripts(project_id: int, db: Session = Depends(get_db)) -> list[ScriptRead]:
    service = ScriptService(db)
    return service.list_scripts(project_id)


@router.post("/projects/{project_id}/scripts", response_model=ScriptRead, status_code=status.HTTP_201_CREATED)
def create_script(project_id: int, payload: ScriptCreate, db: Session = Depends(get_db)) -> ScriptRead:
    service = ScriptService(db)
    return service.create_script(project_id, payload)


@router.get("/scripts/{script_id}", response_model=ScriptRead)
def get_script(script_id: int, db: Session = Depends(get_db)) -> ScriptRead:
    service = ScriptService(db)
    return service.get_script(script_id)


@router.patch("/scripts/{script_id}", response_model=ScriptRead)
def update_script(script_id: int, payload: ScriptUpdate, db: Session = Depends(get_db)) -> ScriptRead:
    service = ScriptService(db)
    return service.update_script(script_id, payload)


@router.delete("/scripts/{script_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_script(script_id: int, db: Session = Depends(get_db)) -> Response:
    service = ScriptService(db)
    service.delete_script(script_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/scripts/generate", response_model=ScriptGenerateResponse)
def generate_script(payload: ScriptGenerateRequest, db: Session = Depends(get_db)) -> ScriptGenerateResponse:
    service = ScriptService(db)
    return service.generate_draft(payload)


@router.get("/scripts/{script_id}/export")
def export_script(script_id: int, format: str = Query(default="json"), db: Session = Depends(get_db)):
    service = ScriptService(db)
    exported = service.export_script(script_id, format)
    if format == "markdown":
        return PlainTextResponse(content=exported, media_type="text/markdown; charset=utf-8")
    return JSONResponse(content=exported)

File: apps/frontend/src/services/scripts.ts
import { API_BASE_URL, http, readErrorMessage } from '@/services/http'
import type {
  ScriptCreate,
  ScriptGenerateRequest,
  ScriptGenerateResponse,
  ScriptRead,
  ScriptUpdate,
} from '@/schemas/script'

export function fetchScripts(projectId: number) {
  return http<ScriptRead[]>(`/api/projects/${projectId}/scripts`)
}

export function createScript(projectId: number, payload: ScriptCreate) {
  return http<ScriptRead>(`/api/projects/${projectId}/scripts`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function updateScript(scriptId: number, payload: ScriptUpdate) {
  return http<ScriptRead>(`/api/scripts/${scriptId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  })
}

export async function deleteScript(scriptId: number) {
  const response = await fetch(`${API_BASE_URL}/api/scripts/${scriptId}`, {
    method: 'DELETE',
  })

  if (!response.ok) {
    throw new Error(await readErrorMessage(response))
  }
}

export function generateScriptDraft(payload: ScriptGenerateRequest) {
  return http<ScriptGenerateResponse>('/api/scripts/generate', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function exportScript(scriptId: number, format: 'json' | 'markdown') {
  const response = await fetch(`${API_BASE_URL}/api/scripts/${scriptId}/export?format=${format}`)

  if (!response.ok) {
    throw new Error(await readErrorMessage(response))
  }

  return format === 'json' ? response.json() : response.text()
}
